sdlib/d/gc/tcache.d

```diff
@@ -421,3 +446,76 @@ unittest getCapacity {
 	assert(p5 !is p4);
 	assert(threadCache.getCapacity(p5[0 .. 16000]) == 20480);
 }
+
+unittest extend {
+	auto nonAppendable = threadCache.alloc(100, false);
+
+	// Attempt to extend a non-appendable:
+	assert(!threadCache.extend(nonAppendable[0 .. 100], 1));
+
+	// Make an appendable alloc:
+	auto p0 = threadCache.allocAppendable(100, false);
+	assert(threadCache.getCapacity(p0[0 .. 100]) == 16384);
+
+	// Attempt to extend slices with capacity 0:
+	assert(!threadCache.extend(p0[0 .. 0], 50));
+	assert(!threadCache.extend(p0[0 .. 99], 50));
+	assert(!threadCache.extend(p0[1 .. 99], 50));
+	assert(!threadCache.extend(p0[0 .. 50], 50));
+
+	// Attempt extend with insufficient space:
+	assert(!threadCache.extend(p0[0 .. 100], 16285));
+	assert(!threadCache.extend(p0[50 .. 100], 16285));
+
+	// Valid extend :
+	assert(threadCache.extend(p0[0 .. 100], 50));
+	auto p1 = p0 + 100;
+	assert(threadCache.getCapacity(p1[0 .. 50]) == 16284);
+
+	// Capacity of old slice becomes 0, as there is now data in front of it:
+	assert(threadCache.getCapacity(p0[0 .. 100]) == 0);
+
+	// Capacity of a slice including the original and the extension:
+	assert(threadCache.getCapacity(p0[0 .. 150]) == 16384);
+
+	// Extend the upper half of p1:
+	assert(threadCache.extend(p1[25 .. 50], 100));
+	auto p4 = p1 + 50;
+	assert(threadCache.getCapacity(p4[0 .. 100]) == 16234);
+
+	// Original's capacity becomes 0:
+	assert(threadCache.getCapacity(p1[25 .. 50]) == 0);
+
+	// Capacity of a slice including original and extended:
+	assert(threadCache.getCapacity(p1[25 .. 150]) == 16259);
+
+	// Capacity of earlier slice elongated to cover the extensions :
+	assert(threadCache.getCapacity(p0[0 .. 250]) == 16384);
+
+	// Extend a zero-length slice existing at the start of the free space:
+	assert(threadCache.extend(p0[250 .. 250], 200));
+	auto p5 = p0 + 250;
+	assert(threadCache.getCapacity(p5[0 .. 200]) == 16134);
+
+	// Capacity of the old slice is now 0:
+	assert(threadCache.getCapacity(p0[0 .. 250]) == 0);
+
+	// Capacity of a slice which includes the original and the extension:
+	assert(threadCache.getCapacity(p0[0 .. 450]) == 16384);
+
+	// Extend so as to fill up all but one byte of free space:
+	assert(threadCache.extend(p0[0 .. 450], 15933));
+	auto p6 = p0 + 450;
+	assert(threadCache.getCapacity(p0[16383 .. 16383]) == 1);
+
+	// // Extend, filling up last byte of free space:
+	assert(threadCache.extend(p0[16383 .. 16383], 1));
+	auto p7 = p0 + 16383;
+	assert(threadCache.getCapacity(p0[0 .. 16384]) == 16384);
+
+	// Attempt to extend, but we're full:
+	assert(!threadCache.extend(p0[0 .. 16384], 1));
+
+	// Extend by length zero still works, though:
+	assert(threadCache.extend(p0[0 .. 16384], 0));
```

Comment:
You should probably sprinkly a couple of extent by 0, to check it works in all circumstances.

Reply:
See https://github.com/snazzy-d/sdc/pull/286/files#diff-0ab2e00d059b668e370db472e12b2b5c427878d8fdffae5186290df9b075df88R380-R388